src/contracts/CrossChainReceiver.sol

```diff
@@ -159,10 +159,21 @@ contract CrossChainReceiver is OwnableWithGuardian, ICrossChainReceiver {
       Errors.CHAIN_ID_MISMATCH
     );
     bytes32 envelopeId = transaction.getEnvelopeId();
-    // if envelope was confirmed before, just return
-    if (_envelopesState[envelopeId] != EnvelopeState.None) return;
 
     bytes32 transactionId = TransactionUtils.getId(encodedTransaction);
+
+    // if envelope was confirmed before, just return
+    if (_envelopesState[envelopeId] != EnvelopeState.None) {
```

Comment:
Not entirely sure if maybe it would be better to just remove this conditional entirely and let the tx enter into the next conditional, where we would just need to put an extra condition to not redeliver message and thats it. 
This would make it so that confirmations get increased if needed, and received tx recorded. It would make it a bit more expensive, but in the end the saved state would be better i think.